salt/modules/x509.py

```diff
@@ -1243,10 +1255,23 @@ def create_csr(path=None, text=False, **kwargs):
 
     csr = M2Crypto.X509.Request()
     subject = csr.get_subject()
+
+    for prop, default in six.iteritems(CERT_DEFAULTS):
+        if prop not in kwargs:
+            kwargs[prop] = default
+
     csr.set_version(kwargs['version'] - 1)
 
+    if 'private_key' not in kwargs and 'public_key' in kwargs:
+        kwargs['private_key'] = kwargs['public_key']
+        log.warning("OpenSSL no longer allows working with non-signed CSRs. A private_key must be specified. Attempting to use public_key as private_key")
+
+    if 'private_key' not in kwargs not in kwargs:
```

Comment:
Previous versions of the OpenSSL API would allow working with CSRs that were not signed. I now suspect that this was not intended behavior, but because it worked I did not require the private key to generate a CSR, only the public key. Now because the CSR needs to be signed the private key must be included, which is why the `private_key` kwarg was added to the module.

Throughout this module anywhere that a public key is specified, a private key can be passed in in. This works because the public key can always be derived easily from the private key, and allowing this means that users of the module have fewer files to manage on disk and keep track of.

Because of that it is quite possible that the user actually provided a private key via the `public_key` kwarg. Instead of just erroring out and requiring the user to rewrite their states, I'll first try and see if the `public_key` kwarg actually does contain a private key, and issue a warning so that they know they need to update their states.